packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -126,96 +172,242 @@ contract MixinExchangeCore is
-        if (order.senderAddress != address(0)) {
-            require(
-                order.senderAddress == msg.sender,
-                INVALID_SENDER
-            );
-        }
-        
         // Validate transaction signed by maker
         address makerAddress = getCurrentContextAddress();
         require(
             order.makerAddress == makerAddress,
             INVALID_CONTEXT
         );
-        
-        if (block.timestamp >= order.expirationTimeSeconds) {
-            emit ExchangeError(uint8(Errors.ORDER_EXPIRED), orderHash);
-            return false;
+
+        // Validate sender is allowed to cancel this order
+        if (order.senderAddress != address(0)) {
+            require(
+                order.senderAddress == msg.sender,
+                INVALID_SENDER
+            );
         }
+    }
 
-        if (cancelled[orderHash]) {
-            emit ExchangeError(uint8(Errors.ORDER_CANCELLED), orderHash);
-            return false;
+    /// @dev Updates state with results of cancelling an order.
+    ///      State is only updated if the order is currently fillable.
+    ///      Otherwise, updating state would have no effect.
+    /// @param order that was cancelled.
+    /// @param orderStatus Status of order that was cancelled.
+    /// @param orderHash Hash of order that was cancelled.
+    /// @return stateUpdated Returns true only if state was updated.
+    function updateCancelledState(
+        Order memory order,
+        uint8 orderStatus,
+        bytes32 orderHash)
+        internal
+        returns (bool stateUpdated)
+    {
+        // Ensure order is fillable (otherwise cancelling does nothing)
+        if (orderStatus != uint8(Status.ORDER_FILLABLE)) {
+            emit ExchangeStatus(uint8(orderStatus), orderHash);
+            stateUpdated = false;
+            return stateUpdated;
         }
 
+        // Perform cancel
         cancelled[orderHash] = true;
+        stateUpdated = true;
 
+        // Log cancel
         emit Cancel(
             order.makerAddress,
             order.feeRecipientAddress,
             orderHash,
             order.makerAssetData,
             order.takerAssetData
         );
-        return true;
+
+        return stateUpdated;
+    }
+
+    /// @dev After calling, the order can not be filled anymore.
+    ///      Throws if order is invalid or sender does not have permission to cancel.
+    /// @param order Order to cancel. Order must be Status.FILLABLE.
+    /// @return True if the order state changed to cancelled.
+    ///         False if the order was order was in a valid, but
+    ///               unfillable state (see LibStatus.STATUS for order states)
+    function cancelOrder(Order memory order)
+        public
+        returns (bool)
+    {
+        // Fetch current order status
+        bytes32 orderHash;
+        uint8 orderStatus;
+        uint256 takerAssetFilledAmount;
+        (orderStatus, orderHash, takerAssetFilledAmount) = getOrderInfo(order);
```

Comment:
Since we're not actually using `takerAssetFilledAmount`, we can write `(orderStatus, orderHash, ) = getOrderInfo(order);`